Return-Path: <holger@tux>
Received: from deliver ([unix socket])
	 by tux (Cyrus 2.5.8) with LMTPA;
	 Tue, 05 Jul 2016 19:18:14 +0200
X-Sieve: CMU Sieve 2.4
Path: news.gmane.org!not-for-mail
From: Qu Wenruo <quwenruo@cn.fujitsu.com>
Newsgroups: gmane.comp.file-systems.btrfs
Subject: [PATCH 3/3] btrfs: qgroup: Fix qgroup incorrectness caused by log replay
Date: Tue,  5 Jul 2016 17:32:07 +0800
Lines: 57
Approved: news@gmane.org
Message-ID: <20160705093207.14983-4-quwenruo@cn.fujitsu.com>
References: <20160705093207.14983-1-quwenruo@cn.fujitsu.com>
NNTP-Posting-Host: plane.gmane.org
Mime-Version: 1.0
X-Trace: ger.gmane.org 1467711190 2391 80.91.229.3 (5 Jul 2016 09:33:10 GMT)
X-Complaints-To: usenet@ger.gmane.org
NNTP-Posting-Date: Tue, 5 Jul 2016 09:33:10 +0000 (UTC)
To: linux-btrfs@vger.kernel.org
Original-X-From: linux-btrfs-owner@vger.kernel.org Tue Jul 05 11:33:09 2016
Envelope-to: gcfb-btrfs-devel-moved1-2@plane.gmane.org
Original-Received: from vger.kernel.org ([209.132.180.67])
	by plane.gmane.org with esmtp (Exim 4.69)
	(envelope-from <linux-btrfs-owner@vger.kernel.org>)
	id 1bKMj6-0004Xm-ES
	for gcfb-btrfs-devel-moved1-2@plane.gmane.org; Tue, 05 Jul 2016 11:33:08 +0200
Original-Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
	id S1754348AbcGEJdE (ORCPT
	<rfc822;gcfb-btrfs-devel-moved1-2@m.gmane.org>);
	Tue, 5 Jul 2016 05:33:04 -0400
Original-Received: from cn.fujitsu.com ([222.73.24.84]:51306 "EHLO song.cn.fujitsu.com"
	rhost-flags-OK-FAIL-OK-OK) by vger.kernel.org with ESMTP
	id S1751150AbcGEJcW (ORCPT <rfc822;linux-btrfs@vger.kernel.org>);
	Tue, 5 Jul 2016 05:32:22 -0400
X-IronPort-AV: E=Sophos;i="5.20,367,1444665600"; 
   d="scan'208";a="656056"
Original-Received: from unknown (HELO cn.fujitsu.com) ([10.167.250.3])
  by song.cn.fujitsu.com with ESMTP; 05 Jul 2016 17:32:12 +0800
Original-Received: from localhost.localdomain (unknown [10.167.226.34])
	by cn.fujitsu.com (Postfix) with ESMTP id A11D04056407
	for <linux-btrfs@vger.kernel.org>; Tue,  5 Jul 2016 17:32:10 +0800 (CST)
X-Mailer: git-send-email 2.9.0
In-Reply-To: <20160705093207.14983-1-quwenruo@cn.fujitsu.com>
X-yoursite-MailScanner-ID: A11D04056407.A7FD6
X-yoursite-MailScanner: Found to be clean
X-yoursite-MailScanner-From: quwenruo@cn.fujitsu.com
X-Spam-Status: No
Original-Sender: linux-btrfs-owner@vger.kernel.org
Precedence: bulk
List-ID: <linux-btrfs.vger.kernel.org>
X-Mailing-List: linux-btrfs@vger.kernel.org
Xref: news.gmane.org gmane.comp.file-systems.btrfs:58003
Archived-At: <http://permalink.gmane.org/gmane.comp.file-systems.btrfs/58003>

When doing log replay at mount time(after power loss), qgroup will leak
numbers of replayed data extents.

The cause is almost the same of balance.
So fix it by manually informing qgroup for owner changed extents.

The bug can be detected by btrfs/119 test case.

Signed-off-by: Qu Wenruo <quwenruo@cn.fujitsu.com>
---
 fs/btrfs/tree-log.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/fs/btrfs/tree-log.c b/fs/btrfs/tree-log.c
index 8ab1dc6..a322aa5 100644
--- a/fs/btrfs/tree-log.c
+++ b/fs/btrfs/tree-log.c
@@ -27,6 +27,7 @@
 #include "backref.h"
 #include "hash.h"
 #include "compression.h"
+#include "qgroup.h"
 
 /* magic values for the inode_only field in btrfs_log_inode:
  *
@@ -680,6 +681,21 @@ static noinline int replay_one_extent(struct btrfs_trans_handle *trans,
 		ins.type = BTRFS_EXTENT_ITEM_KEY;
 		offset = key->offset - btrfs_file_extent_offset(eb, item);
 
+		/*
+		 * Manually record dirty extent, as here we did a shallow
+		 * file extent item copy and skip normal backref update,
+		 * but modify extent tree all by ourselves.
+		 * So need to manually record dirty extent for qgroup,
+		 * as the owner of the file extent changed from log tree
+		 * (doesn't affect qgroup) to fs/file tree(affects qgroup)
+		 */
+		ret = btrfs_qgroup_record_dirty_extent(trans, root->fs_info,
+				btrfs_file_extent_disk_bytenr(eb, item),
+				btrfs_file_extent_disk_num_bytes(eb, item),
+				GFP_NOFS);
+		if (ret < 0)
+			goto out;
+
 		if (ins.objectid > 0) {
 			u64 csum_start;
 			u64 csum_end;
-- 
2.9.0



--
To unsubscribe from this list: send the line "unsubscribe linux-btrfs" in
the body of a message to majordomo@vger.kernel.org
More majordomo info at  http://vger.kernel.org/majordomo-info.html

